File: ros_cv/ros_cv/cam_sub.py
'''
ROS2 OpenCCV Image Subscriber
ROS2 토픽으로 발행된 영상을 수신, 화면에 표시
'''

import rclpy
from rclpy.node import Node
from sensor_msgs.msg import Image
from cv_bridge import CvBridge
import cv2

class ImageSubscriber(Node):
    """
    ImageSubscriber Class
    ROS2 Image 메시지를 구독하여 OpenCV 창에 표시
    """

    def __init__(self):
        super().__init__('image_subscriber')

        self.subscription = self.create_subscription(
            Image,
            'video_frames',
            self.listener_callback,
            10
        )

        self.subscription
        self.br = CvBridge()
        self.get_logger().info('Image Subscriber Node Started')
    
    def listener_callback(self, data):
        self.get_logger().info('Receiving video frame')
        current_frame = self.br.imgmsg_to_cv2(data)
        cv2.imshow('camera', current_frame)
        cv2.waitKey(1)

def main(args=None):
    rclpy.init(args=args)
    image_subscriber = ImageSubscriber()

    try:
        rclpy.spin(image_subscriber)
    except KeyboardInterrupt:
        pass
    finally:
        cv2.destroyAllWindows()
        image_subscriber.destroy_node()
        rclpy.shutdown()

        
if __name__ == '__main__':
    main()

File: ros_cv_c/src/cam_sub.cpp
#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/image.hpp>
#include <cv_bridge/cv_bridge.h>
#include <opencv2/opencv.hpp>
#include <opencv2/highgui/highgui.hpp>

class ImageSubscriber : public rclcpp::Node
{
public:
    ImageSubscriber() : Node("image_subscriber")
    {
        cv::namedWindow(OPENCV_WINDOW, cv::WINDOW_AUTOSIZE);
        subscription_ = this->create_subscription<sensor_msgs::msg::Image>(
            "camera/image",
            10,
            std::bind(&ImageSubscriber::image_callback, this, std::placeholders::_1)
        );
        RCLCPP::INFO(this->get_logger(),
                "Image Subscriber Node started, waiting for images on 'camera/image' topic");
    }

    ~ImageSubscriber()
    {
        cv::destroyWindow(OPENCV_WINDOW);
    }
private:
    void image_callback(const sensor_msgs::msg::Image::SharedPtr msg)
    {
        cv_bridge::CvImagePtr cv_ptr;

        try
        {
            cv_ptr = cv_bridge::toCvCopy(msg, msg->encoding);
        }
        catch (cv_bridge::Exception& e)
        {
            RCLCPP::ERROR(this->get_logger(), "cv_bridge exception: %s", e.what());
            return;
        }

        cv::circle(
            cv_ptr->image,
            cv::Point(cv_ptr->image.cols / 2, cv_ptr->image.rows / 2),
            20,
            CV_RGB(255, 0, 0),
            2
        );

        cv::imshow(OPENCV_WINDOW, cv_ptr->image);

        cv::waitKey(3);

        RCLCPP_INFO(this->get_logger(), "Received image. Time stamp: %f",
                    this->now().seconds());
    }

    rclcpp::Subscription<sensor_msgs::msg::Image>::SharedPtr subscription_;
    const std::string OPENCV_WINDOW = "Webcam Feed";


};

int main(int argc, char * argv[])
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<ImageSubscriber>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}